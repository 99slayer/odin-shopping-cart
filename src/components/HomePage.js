import React, { useState, useEffect } from "react";
import "../styles/HomePage.css";

export const HomePage = () => {
  return (
    <div id="home-page">
      <ProductDisplay />
      <Reviews />
      <About />
    </div>
  );
};

const ProductDisplay = () => {
  return (
    <div id="product-display">
      <img alt="" src="toaster-1.jpg"></img>
      <p>
        Introducing the next leap forward in toaster technology! The Toast
        Master 5000 combines turbo charged industrial strength toast chambers
        with varied and sophisticated user settings. Toasting the perfect slice
        has never been easier!
      </p>
    </div>
  );
};

const reviewList = [
  '"I look forward to breakfast every morning now!"',
  "\"I've spent years looking for the perfect toaster, and now I've found it.\"",
  '"These toaster are the best! I don\'t go anywhere without one!"',
  '"Toaster Warehouse saved my marriage!"',
  '"These make the perfect Valentine\'s Day gift. Nothing is more romantic than toast!"',
  '"I\'ve been let down by alot of toasters so I was skeptical at first, but boy was I wrong!"',
  '"Toaster Warehouse are pioneers in the field of toasters."',
  '"When I need a toaster, I know who to call."',
  '"These toasters can\'t be beat!"',
  '"Only Toaster Warehouse brand toasters touch my counter top."',
  '"Busting out one of these beauties at a family gathering makes you the life of the party."',
  '"I thought I knew what was possible with toast. But my toast journey was just beginning!"',
  '"A couple slices and you\'ll understand why these guys are the best!"',
];

const currentRevs = [0, 1, 2];

function getRandom(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getUnique(arr) {
  if (arr.length === 0) {
    return getRandom(0, reviewList.length - 1);
  } else {
    let newIndex = getRandom(0, reviewList.length - 1);

    while (arr.includes(newIndex)) {
      newIndex = getRandom(0, reviewList.length - 1);
    }

    return newIndex;
  }
}

const Reviews = () => {
  const [review1, setReview1] = useState(reviewList[0]);
  const [review2, setReview2] = useState(reviewList[1]);
  const [review3, setReview3] = useState(reviewList[2]);

  useEffect(() => {
    setInterval(() => {
      const index = getUnique(currentRevs);
      currentRevs.splice(0, 1, index);
      setReview1(reviewList[index]);
    }, 5000);

    setInterval(() => {
      const index = getUnique(currentRevs);
      currentRevs.splice(1, 1, index);
      setReview2(reviewList[index]);
    }, 5000);

    setInterval(() => {
      const index = getUnique(currentRevs);
      currentRevs.splice(2, 1, index);
      setReview3(reviewList[index]);
    }, 5000);
  }, []);

  return (
    <div id="reviews">
      <h2 id="review-header" aria-label="review header">
        SATISFIED CUSTOMER REVIEWS
      </h2>
      <ul id="review-cont">
        <li id="review-column-1" className="review-column">
          {review1}
        </li>
        <li id="review-column-2" className="review-column">
          {review2}
        </li>
        <li id="review-column-3" className="review-column">
          {review3}
        </li>
      </ul>
    </div>
  );
};

const About = () => {
  return (
    <div id="about">
      <h2 id="about-header">ABOUT US</h2>
      <p id="about-text">
        Toaster Warehouse was first founded in 1892 in a small town called Loaf
        Shire. It's founder John Goodenbread had a saying,
        <strong id="motto">
          {" 'Reasonably quality products at reasonably affordable prices.' "}
        </strong>
        This motto continues to guide our company all these years later. From
        our warehouse to your counter, nothing brings the family together like
        toast!
      </p>
    </div>
  );
};
