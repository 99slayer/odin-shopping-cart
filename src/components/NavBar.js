import React from "react";
import '../styles/NavBar.css';
import { Link } from "react-router-dom";

export const NavBar = () => {
  return (
    <div id="nav-bar">
      <h1 id="heading">SHOPPING PROJECT</h1>
      <div id="nav-links">
        <Link to='/homepage'>HOME</Link >
        <Link to='/shoppage'>SHOP</Link>
      </div>
    </div>
  );
};