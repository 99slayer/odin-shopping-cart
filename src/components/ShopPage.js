import React from "react";
import { useOutletContext } from "react-router-dom";
import "../styles/ShopPage.css";

const order = [];

// Manages order array so cart items can be rendered in the order that they are added or removed.
const changeOrder = (list) => {
  for (const product in list) {
    if (order.includes(list[product].name) && list[product].count < 1) {
      const index = order.indexOf(list[product].name);
      order.splice(index, 1);
    } else if (order.includes(list[product].name)) {
      continue;
    } else if (list[product].count < 1) {
      continue;
    } else {
      order.push(list[product].name);
    }
  }
};

export const ShopPage = () => {
  const [productList, setProductList] = useOutletContext();

  const incrementProduct = (productNum) => {
    if (productList[productNum].count === 999) {
      return;
    }

    setProductList((prevList) => ({
      ...prevList,
      [productNum]: {
        ...prevList[productNum],
        count: parseInt(prevList[productNum].count + 1),
      },
    }));
  };

  const decrementProduct = (productNum) => {
    if (productList[productNum].count === 0) {
      return;
    }

    setProductList((prevList) => ({
      ...prevList,
      [productNum]: {
        ...prevList[productNum],
        count: parseInt(prevList[productNum].count - 1),
      },
    }));
  };

  const removeProduct = (productNum) => {
    setProductList((prevList) => ({
      ...prevList,
      [productNum]: { ...prevList[productNum], count: 0 },
    }));
  };

  // For dealing with user input.
  const handleProductInput = (productNum, e) => {
    if (e.target.value.length > 3 && e.target.key !== "Backspace") {
      return;
    }

    if (e.target.value[0] === "0") {
      const arr = e.target.value.split("");
      arr.splice(0, 1);
      e.target.value = arr.join("");
    }

    if (e.target.value === "") {
      e.target.value = "0";
    }

    setProductList((prevList) => ({
      ...prevList,
      [productNum]: {
        ...prevList[productNum],
        count: parseInt(e.target.value),
      },
    }));
  };

  const getTotal = (list) => {
    let total = 0;

    for (const product in list) {
      total += list[product].count * list[product].cost;
    }

    return total > 0 ? total.toFixed(2) : "";
  };

  return (
    <div id="shop-page">
      <ShoppingGrid
        increment={incrementProduct}
        decrement={decrementProduct}
        inputFunc={handleProductInput}
        list={productList}
      />
      <ShoppingCart
        increment={incrementProduct}
        decrement={decrementProduct}
        removeProduct={removeProduct}
        inputFunc={handleProductInput}
        totalFunc={getTotal}
        list={productList}
      />
    </div>
  );
};

const ShoppingGrid = (props) => {
  const { increment, decrement, inputFunc, list } = props;

  const renderCards = (obj) => {
    const elements = [];
    let i = 1;

    for (const product in obj) {
      elements.push(
        <ProductCard
          key={i}
          name={obj[product].name}
          cost={obj[product].cost}
          count={obj[product].count}
          image={obj[product].image}
          increment={increment}
          decrement={decrement}
          inputFunc={inputFunc}
          productNum={i}
        />
      );
      i += 1;
    }

    return elements;
  };

  return <div id="shopping-grid">{renderCards(list)}</div>;
};

const ProductCard = (props) => {
  const {
    name,
    cost,
    count,
    image,
    increment,
    decrement,
    inputFunc,
    productNum,
  } = props;

  const handleKeyDown = (e) => {
    const validInputs = [
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "0",
      "Backspace",
    ];

    if (!validInputs.includes(e.key)) {
      e.preventDefault();
    }
  };

  return (
    <div className="product-card">
      <img alt="" src={image} />
      <div className="product-card-infobar">
        <p className="product-card-name">{name}</p>
        <p className="product-card-cost">{"$" + cost.toFixed(2)}</p>
        <div className="product-card-count">
          <svg width={20} height={20}>
            <a onClick={() => decrement(productNum)}>
              <polygon className="decrement-btn" points="3,10 20,0 20,20" />
            </a>
          </svg>
          <input
            type="number"
            value={count}
            maxLength={3}
            onChange={(e) => {
              inputFunc(productNum, e);
            }}
            onKeyDown={handleKeyDown}
          ></input>
          <svg width={20} height={20}>
            <a onClick={() => increment(productNum)}>
              <polygon className="increment-btn" points="17,10 0,20 0,0" />
            </a>
          </svg>
        </div>
      </div>
    </div>
  );
};

const ShoppingCart = (props) => {
  const { increment, decrement, removeProduct, inputFunc, totalFunc, list } =
    props;

  const renderItems = (obj) => {
    changeOrder(list);
    const elements = [];
    let i = 1;

    while (elements.length !== order.length) {
      for (const product in obj) {
        if (obj[product].name !== order[i - 1]) {
          continue;
        } else {
          elements.push(
            <Item
              key={product}
              name={obj[product].name}
              cost={obj[product].cost}
              count={obj[product].count}
              increment={increment}
              decrement={decrement}
              removeProduct={removeProduct}
              inputFunc={inputFunc}
              productNum={product}
            />
          );

          i += 1;
        }
      }
    }

    return elements;
  };

  return (
    <div id="shopping-cart">
      <h2 id="cart-heading">CART</h2>
      <div id="item-cont">{renderItems(list)}</div>
      {/* can the getTotal function be moved to this component? */}
      <div
        id="cart-total"
        className={totalFunc(list) === "" ? "no-total" : ""}
      >{`TOTAL: $${totalFunc(list)}`}</div>
      <button id="checkout-btn">CHECKOUT</button>
    </div>
  );
};

const Item = (props) => {
  const {
    name,
    cost,
    count,
    increment,
    decrement,
    removeProduct,
    inputFunc,
    productNum,
  } = props;

  const handleKeyDown = (e) => {
    const validInputs = [
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "0",
      "Backspace",
    ];

    if (!validInputs.includes(e.key)) {
      e.preventDefault();
    }
  };

  return (
    <div className="item">
      <p className="item-name">{name}</p>
      <p className="item-cost">{"$" + (cost * count).toFixed(2)}</p>
      <div className="item-count-cont">
        <svg width={20} height={20}>
          <a onClick={() => decrement(productNum)}>
            <polygon className="decrement-btn" points="3,10 20,0 20,20" />
          </a>
        </svg>
        <input
          type="number"
          value={count}
          maxLength={3}
          onChange={(e) => {
            inputFunc(productNum, e);
          }}
          onKeyDown={handleKeyDown}
        ></input>
        <svg width={20} height={20}>
          <a onClick={() => increment(productNum)}>
            <polygon className="increment-btn" points="17,10 0,20 0,0" />
          </a>
        </svg>
      </div>
      <button
        type="button"
        className="item-delete-btn"
        onClick={() => removeProduct(productNum)}
      >
        X
      </button>
    </div>
  );
};
