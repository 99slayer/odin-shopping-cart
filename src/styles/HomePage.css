ul,
li {
  list-style: none;
}
#home-page {
  flex: 1;
  height: 100vh;
  padding: 10px;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  gap: 10px;
}
#product-display {
  flex: 1;
  padding: 10px;
  display: flex;
  justify-content: flex start;
  gap: 10px;
  box-shadow: rgba(0, 0, 0, 0.19) 0px 10px 20px, rgba(0, 0, 0, 0.23) 0px 6px 6px;
  border-radius: 6px;
}
#product-display img {
  width: 300px;
  height: 300px;
  border: 1px solid rgb(196, 196, 196);
  border-radius: 6px;
}
#product-display p {
  flex: 1;
  padding: 10px;
  font-size: 2.6em;
  border: 1px solid rgb(196, 196, 196);
  border-radius: 6px;
  background-color: white;
}
#reviews {
  flex: 1;
  padding: 10px;
  margin-top: 20px;
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  gap: 10px;
  box-shadow: rgba(0, 0, 0, 0.19) 0px 10px 20px, rgba(0, 0, 0, 0.23) 0px 6px 6px;
  border-radius: 6px;
}
#review-header {
  padding: 6px 0px 6px 10px;
  font-size: 2.2em;
  border: 1px solid rgb(196, 196, 196);
  border-radius: 6px;
  background-color: white;
}
#review-cont {
  min-height: 190px;
  display: grid;
  grid-template-rows: 1fr;
  grid-template-columns: 1fr 1fr 1fr;
  gap: 10px;
}
.review-column {
  padding: 10px;
  font-size: 2em;
  border: 1px solid rgb(196, 196, 196);
  border-radius: 6px;
  background-color: white;
}
#about {
  flex: 1;
  padding: 10px;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  gap: 10px;
  box-shadow: rgba(0, 0, 0, 0.19) 0px 10px 20px, rgba(0, 0, 0, 0.23) 0px 6px 6px;
  border-radius: 6px;
}
#about-header {
  padding: 6px 0px 6px 10px;
  font-size: 2.2em;
  border: 1px solid rgb(196, 196, 196);
  border-radius: 6px;
  background-color: white;
}
#about-text {
  padding: 10px;
  font-size: 2.2em;
  border: 1px solid rgb(196, 196, 196);
  border-radius: 6px;
  background-color: white;
}
#motto {
  font-size: 1.2em;
}
