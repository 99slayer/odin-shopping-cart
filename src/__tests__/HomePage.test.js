import React from "react";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import { HomePage } from "../components/HomePage";

describe("HomePage component", () => {
  it("matches snapshot", () => {
    const homePageComponent = renderer.create(<HomePage />).toJSON();
    expect(homePageComponent).toMatchSnapshot();
  });
});
