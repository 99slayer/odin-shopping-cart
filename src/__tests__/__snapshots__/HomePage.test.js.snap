// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage component matches snapshot 1`] = `
<div
  id="home-page"
>
  <div
    id="product-display"
  >
    <img
      alt=""
      src="toaster-1.jpg"
    />
    <p>
      Introducing the next leap forward in toaster technology! The Toast Master 5000 combines turbo charged industrial strength toast chambers with varied and sophisticated user settings. Toasting the perfect slice has never been easier!
    </p>
  </div>
  <div
    id="reviews"
  >
    <h2
      aria-label="review header"
      id="review-header"
    >
      SATISFIED CUSTOMER REVIEWS
    </h2>
    <ul
      id="review-cont"
    >
      <li
        className="review-column"
        id="review-column-1"
      >
        "I look forward to breakfast every morning now!"
      </li>
      <li
        className="review-column"
        id="review-column-2"
      >
        "I've spent years looking for the perfect toaster, and now I've found it."
      </li>
      <li
        className="review-column"
        id="review-column-3"
      >
        "These toaster are the best! I don't go anywhere without one!"
      </li>
    </ul>
  </div>
  <div
    id="about"
  >
    <h2
      id="about-header"
    >
      ABOUT US
    </h2>
    <p
      id="about-text"
    >
      Toaster Warehouse was first founded in 1892 in a small town called Loaf Shire. It's founder John Goodenbread had a saying,
      <strong
        id="motto"
      >
         'Reasonably quality products at reasonably affordable prices.' 
      </strong>
      This motto continues to guide our company all these years later. From our warehouse to your counter, nothing brings the family together like toast!
    </p>
  </div>
</div>
`;
